File: c++/src/whenwords.cpp
#include "whenwords.h"
#include <string>
#include <sstream>
#include <vector>
#include <regex>
#include <cmath>
#include <iomanip>
#include <ctime>

namespace whenwords {

// Helper: Normalize timestamp to int64_t
int64_t normalize_timestamp(const std::string& timestamp) {
    // Try to parse as integer
    try {
        size_t pos;
        int64_t value = std::stoll(timestamp, &pos);
        if (pos == timestamp.size()) {
            return value;
        }
    } catch (...) {}
    
    // Try to parse as double
    try {
        size_t pos;
        double value = std::stod(timestamp, &pos);
        if (pos == timestamp.size()) {
            return static_cast<int64_t>(value);
        }
    } catch (...) {}
    
    // Try to parse as ISO 8601
    // This is a simplified parser - real implementation would be more robust
    std::regex iso8601_regex(R"(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})?)");
    if (std::regex_match(timestamp, iso8601_regex)) {
        // For simplicity, we'll just throw an error for ISO 8601 for now
        // In a real implementation, we'd parse this properly
        throw ParseError("ISO 8601 parsing not implemented");
    }
    
    throw ParseError("Invalid timestamp format");
}

int64_t normalize_timestamp(int64_t timestamp) {
    return timestamp;
}

int64_t normalize_timestamp(double timestamp) {
    return static_cast<int64_t>(timestamp);
}

// Helper: Format time ago string
std::string format_timeago(int64_t diff) {
    bool future = diff < 0;
    diff = std::abs(diff);
    
    if (diff < 45) {
        return "just now";
    } else if (diff < 90) {
        return future ? "in 1 minute" : "1 minute ago";
    } else if (diff < 45 * 60) {
        int minutes = static_cast<int>(std::round(diff / 60.0));
        return future ? "in " + std::to_string(minutes) + " minutes" : 
               std::to_string(minutes) + " minutes ago";
    } else if (diff < 90 * 60) {
        return future ? "in 1 hour" : "1 hour ago";
    } else if (diff < 22 * 3600) {
        int hours = static_cast<int>(std::round(diff / 3600.0));
        return future ? "in " + std::to_string(hours) + " hours" : 
               std::to_string(hours) + " hours ago";
    } else if (diff < 36 * 3600) {
        return future ? "in 1 day" : "1 day ago";
    } else if (diff < 26 * 86400) {
        int days = static_cast<int>(std::round(diff / 86400.0));
        return future ? "in " + std::to_string(days) + " days" : 
               std::to_string(days) + " days ago";
    } else if (diff < 46 * 86400) {
        return future ? "in 1 month" : "1 month ago";
    } else if (diff < 320 * 86400) {
        int months = static_cast<int>(std::round(diff / (30.0 * 86400)));
        return future ? "in " + std::to_string(months) + " months" : 
               std::to_string(months) + " months ago";
    } else if (diff < 548 * 86400) {
        return future ? "in 1 year" : "1 year ago";
    } else {
        int years = static_cast<int>(std::round(diff / (365.0 * 86400)));
        return future ? "in " + std::to_string(years) + " years" : 
               std::to_string(years) + " years ago";
    }
}

// timeago implementations
std::string timeago(const std::string& timestamp, const std::string& reference) {
    int64_t ts = normalize_timestamp(timestamp);
    int64_t ref = reference.empty() ? ts : normalize_timestamp(reference);
    int64_t diff = ref - ts;
    return format_timeago(diff);
}

std::string timeago(int64_t timestamp, int64_t reference) {
    int64_t ref = (reference == -1) ? timestamp : reference;
    int64_t diff = ref - timestamp;
    return format_timeago(diff);
}

std::string timeago(double timestamp, double reference) {
    int64_t ts = normalize_timestamp(timestamp);
    int64_t ref = (reference == -1) ? ts : normalize_timestamp(reference);
    int64_t diff = ref - ts;
    return format_timeago(diff);
}

// duration implementation
std::string duration(double seconds, const DurationOptions& options) {
    if (seconds < 0) {
        throw ParseError("Duration cannot be negative");
    }
    if (std::isnan(seconds) || std::isinf(seconds)) {
        throw ParseError("Invalid duration value");
    }
    
    struct Unit {
        const char* name;
        const char* compact;
        double seconds;
    } units[] = {
        {"year", "y", 365 * 86400},
        {"month", "m", 30 * 86400},
        {"day", "d", 86400},
        {"hour", "h", 3600},
        {"minute", "m", 60},
        {"second", "s", 1}
    };
    
    std::vector<std::pair<double, const Unit*>> parts;
    double remaining = seconds;
    
    for (auto& unit : units) {
        if (remaining >= unit.seconds || unit.seconds == 1) {
            double value = remaining / unit.seconds;
            parts.emplace_back(value, &unit);
            remaining = std::fmod(remaining, unit.seconds);
        }
    }
    
    // Apply max_units
    if (parts.size() > static_cast<size_t>(options.max_units)) {
        parts.resize(options.max_units);
        // Round the last part
        if (!parts.empty()) {
            parts.back().first = std::round(parts.back().first);
        }
    }
    
    // Build the result
    std::stringstream ss;
    bool first = true;
    
    for (auto& [value, unit] : parts) {
        int count = static_cast<int>(value);
        if (count == 0) continue;
        
        if (!first) {
            if (options.compact) {
                ss << " ";
            } else {
                ss << ", ";
            }
        }
        
        ss << count;
        if (options.compact) {
            ss << unit->compact;
        } else {
            ss << " " << unit->name;
            if (count != 1) ss << "s";
        }
        
        first = false;
    }
    
    if (ss.str().empty()) {
        return options.compact ? "0s" : "0 seconds";
    }
    
    return ss.str();
}

// parse_duration implementation
int64_t parse_duration(const std::string& input) {
    if (input.empty()) {
        throw ParseError("Empty duration string");
    }
    
    // Check for negative
    if (!input.empty() && input[0] == '-') {
        throw ParseError("Negative duration not allowed");
    }
    
    // Check for colon notation
    if (input.find(':') != std::string::npos) {
        std::vector<std::string> parts;
        std::stringstream ss(input);
        std::string part;
        while (std::getline(ss, part, ':')) {
            parts.push_back(part);
        }
        
        if (parts.size() == 2) {
            // h:mm
            int hours = std::stoi(parts[0]);
            int minutes = std::stoi(parts[1]);
            return hours * 3600 + minutes * 60;
        } else if (parts.size() == 3) {
            // h:mm:ss
            int hours = std::stoi(parts[0]);
            int minutes = std::stoi(parts[1]);
            int seconds = std::stoi(parts[2]);
            return hours * 3600 + minutes * 60 + seconds;
        }
    }
    
    // Regex pattern for duration parts (with optional whitespace)
    std::regex pattern(R"((\d+(?:\.\d+)?)\s*([a-zA-Z]+))");
    std::smatch matches;
    std::string::const_iterator search_start(input.cbegin());
    
    int64_t total = 0;
    bool found = false;
    
    while (std::regex_search(search_start, input.cend(), matches, pattern)) {
        found = true;
        double value = std::stod(matches[1]);
        std::string unit = matches[2];
        
        // Normalize unit to lowercase
        std::transform(unit.begin(), unit.end(), unit.begin(), ::tolower);
        
        // Debug output
        // std::cerr << "Match: " << matches[1] << " " << unit << std::endl;
        
        if (unit == "s" || unit == "sec" || unit == "secs" || unit == "second" || unit == "seconds") {
            total += static_cast<int64_t>(value);
        } else if (unit == "m" || unit == "min" || unit == "mins" || unit == "minute" || unit == "minutes") {
            total += static_cast<int64_t>(value * 60);
        } else if (unit == "h" || unit == "hr" || unit == "hrs" || unit == "hour" || unit == "hours") {
            total += static_cast<int64_t>(value * 3600);
        } else if (unit == "d" || unit == "day" || unit == "days") {
            total += static_cast<int64_t>(value * 86400);
        } else if (unit == "w" || unit == "wk" || unit == "wks" || unit == "week" || unit == "weeks") {
            total += static_cast<int64_t>(value * 604800);
        }
        
        search_start = matches.suffix().first;
    }
    
    if (!found) {
        throw ParseError("No parseable duration units found");
    }
    
    return total;
}

// Helper: Convert timestamp to tm struct in UTC
std::tm timestamp_to_tm(int64_t timestamp) {
    std::tm tm = {};
    std::time_t t = static_cast<std::time_t>(timestamp);
    gmtime_r(&t, &tm);
    return tm;
}

// Helper: Format date
std::string format_date(const std::tm& tm, bool include_year = true) {
    const char* months[] = {
        "January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December"
    };
    
    std::stringstream ss;
    ss << months[tm.tm_mon] << " " << (tm.tm_mday);
    if (include_year) {
        ss << ", " << (tm.tm_year + 1900);
    }
    return ss.str();
}

// Helper: Get weekday name
std::string get_weekday_name(int weekday) {
    const char* weekdays[] = {
        "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
    };
    return weekdays[weekday];
}

// human_date implementation
std::string human_date(const std::string& timestamp, const std::string& reference) {
    int64_t ts = normalize_timestamp(timestamp);
    int64_t ref = reference.empty() ? ts : normalize_timestamp(reference);
    return human_date(ts, ref);
}

std::string human_date(int64_t timestamp, int64_t reference) {
    int64_t ref = (reference == -1) ? timestamp : reference;
    
    std::tm ts_tm = timestamp_to_tm(timestamp);
    std::tm ref_tm = timestamp_to_tm(ref);
    
    // Check if same day
    if (ts_tm.tm_year == ref_tm.tm_year && 
        ts_tm.tm_mon == ref_tm.tm_mon && 
        ts_tm.tm_mday == ref_tm.tm_mday) {
        return "Today";
    }
    
    // Check if yesterday
    std::tm yesterday = ref_tm;
    yesterday.tm_mday--;
    mktime(&yesterday); // Normalize
    if (ts_tm.tm_year == yesterday.tm_year && 
        ts_tm.tm_mon == yesterday.tm_mon && 
        ts_tm.tm_mday == yesterday.tm_mday) {
        return "Yesterday";
    }
    
    // Check if tomorrow
    std::tm tomorrow = ref_tm;
    tomorrow.tm_mday++;
    mktime(&tomorrow); // Normalize
    if (ts_tm.tm_year == tomorrow.tm_year && 
        ts_tm.tm_mon == tomorrow.tm_mon && 
        ts_tm.tm_mday == tomorrow.tm_mday) {
        return "Tomorrow";
    }
    
    // Check if within past 7 days
    int64_t diff_days = (ref - timestamp) / 86400;
    if (diff_days > 0 && diff_days < 7) {
        return "Last " + get_weekday_name(ts_tm.tm_wday);
    }
    
    // Check if within next 7 days
    diff_days = (timestamp - ref) / 86400;
    if (diff_days > 0 && diff_days < 7) {
        return "This " + get_weekday_name(ts_tm.tm_wday);
    }
    
    // Check if same year
    if (ts_tm.tm_year == ref_tm.tm_year) {
        return format_date(ts_tm, false);
    }
    
    // Different year
    return format_date(ts_tm, true);
}

std::string human_date(double timestamp, double reference) {
    int64_t ts = normalize_timestamp(timestamp);
    int64_t ref = (reference == -1) ? ts : normalize_timestamp(reference);
    return human_date(ts, ref);
}

// date_range implementation
std::string date_range(const std::string& start, const std::string& end) {
    int64_t s = normalize_timestamp(start);
    int64_t e = normalize_timestamp(end);
    return date_range(s, e);
}

std::string date_range(int64_t start, int64_t end) {
    // Swap if start > end
    if (start > end) {
        std::swap(start, end);
    }
    
    std::tm start_tm = timestamp_to_tm(start);
    std::tm end_tm = timestamp_to_tm(end);
    
    // Same day
    if (start_tm.tm_year == end_tm.tm_year && 
        start_tm.tm_mon == end_tm.tm_mon && 
        start_tm.tm_mday == end_tm.tm_mday) {
        return format_date(start_tm, true);
    }
    
    // Same month
    if (start_tm.tm_year == end_tm.tm_year && 
        start_tm.tm_mon == end_tm.tm_mon) {
        const char* months[] = {
            "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
        };
        std::stringstream ss;
        ss << months[start_tm.tm_mon] << " " << start_tm.tm_mday << "–" << end_tm.tm_mday;
        ss << ", " << (start_tm.tm_year + 1900);
        return ss.str();
    }
    
    // Same year
    if (start_tm.tm_year == end_tm.tm_year) {
        const char* months[] = {
            "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
        };
        std::stringstream ss;
        ss << months[start_tm.tm_mon] << " " << start_tm.tm_mday << " – ";
        ss << months[end_tm.tm_mon] << " " << end_tm.tm_mday;
        ss << ", " << (start_tm.tm_year + 1900);
        return ss.str();
    }
    
    // Different years
    return format_date(start_tm, true) + " – " + format_date(end_tm, true);
}

std::string date_range(double start, double end) {
    int64_t s = normalize_timestamp(start);
    int64_t e = normalize_timestamp(end);
    return date_range(s, e);
}

} // namespace whenwords

File: java/src/whenwords.py
from datetime import datetime
from typing import Union, Optional, Dict, Any
import math


def _to_timestamp(ts: Union[int, float, str]) -> float:
    if isinstance(ts, str):
        try:
            dt = datetime.fromisoformat(ts.replace('Z', '+00:00'))
            return dt.timestamp()
        except ValueError:
            raise ValueError(f"Invalid timestamp format: {ts}")
    return float(ts)


def _pluralize(n: int, singular: str, plural: Optional[str] = None) -> str:
    if plural is None:
        plural = singular + 's'
    return singular if n == 1 else plural


def timeago(timestamp: Union[int, float, str], reference: Optional[Union[int, float, str]] = None) -> str:
    ts = _to_timestamp(timestamp)
    ref = _to_timestamp(reference) if reference is not None else ts
    
    diff = ref - ts
    is_future = diff < 0
    diff = abs(diff)
    
    if diff < 45:
        return "just now"
    
    def format_result(value: int, unit: str) -> str:
        unit_str = _pluralize(value, unit)
        if is_future:
            return f"in {value} {unit_str}"
        return f"{value} {unit_str} ago"
    
    if diff < 90:
        return format_result(1, "minute")
    
    if diff < 45 * 60:
        minutes = round(diff / 60)
        return format_result(minutes, "minute")
    
    if diff < 90 * 60:
        return format_result(1, "hour")
    
    if diff < 22 * 3600:
        hours = round(diff / 3600)
        return format_result(hours, "hour")
    
    if diff < 36 * 3600:
        return format_result(1, "day")
    
    if diff < 26 * 86400:
        days = round(diff / 86400)
        return format_result(days, "day")
    
    if diff < 46 * 86400:
        return format_result(1, "month")
    
    if diff < 320 * 86400:
        months = round(diff / (30 * 86400))
        return format_result(months, "month")
    
    if diff < 548 * 86400:
        return format_result(1, "year")
    
    years = round(diff / (365 * 86400))
    return format_result(years, "year")


def duration(seconds: Union[int, float], options: Optional[Dict[str, Any]] = None) -> str:
    if seconds < 0 or math.isnan(seconds) or math.isinf(seconds):
        raise ValueError("seconds must be non-negative and finite")
    
    opts = options or {}
    compact = opts.get('compact', False)
    max_units = opts.get('max_units', 2)
    
    if seconds == 0:
        return "0s" if compact else "0 seconds"
    
    remaining = seconds
    units = []
    
    year_seconds = 365 * 86400
    month_seconds = 30 * 86400
    day_seconds = 86400
    hour_seconds = 3600
    minute_seconds = 60
    
    unit_defs = [
        ('year', year_seconds),
        ('month', month_seconds),
        ('day', day_seconds),
        ('hour', hour_seconds),
        ('minute', minute_seconds),
        ('second', 1),
    ]
    
    for name, unit_seconds in unit_defs:
        if remaining >= unit_seconds:
            count = int(remaining // unit_seconds)
            units.append((name, count))
            remaining -= count * unit_seconds
    
    if len(units) > max_units and max_units == 1:
        largest_name = units[0][0]
        for name, unit_seconds in unit_defs:
            if name == largest_name:
                total_in_largest = seconds / unit_seconds
                rounded = int(math.floor(total_in_largest + 0.5))
                units = [(largest_name, rounded)]
                break
    elif len(units) > max_units:
        units = units[:max_units]
    
    if compact:
        parts = [f"{count}{name[0]}" for name, count in units]
        return ' '.join(parts)
    else:
        parts = [f"{count} {_pluralize(count, name)}" for name, count in units]
        return ', '.join(parts)


def parse_duration(string: str) -> float:
    import re
    
    s = string.strip().lower()
    if not s:
        raise ValueError("Empty string")
    
    if s.startswith('-'):
        raise ValueError("Negative duration")
    
    colon_match = re.match(r'^(\d+):(\d+)(?::(\d+))?$', s)
    if colon_match:
        hours = int(colon_match.group(1))
        minutes = int(colon_match.group(2))
        seconds = int(colon_match.group(3)) if colon_match.group(3) else 0
        return hours * 3600 + minutes * 60 + seconds
    
    unit_map = {
        's': 1, 'sec': 1, 'secs': 1, 'second': 1, 'seconds': 1,
        'm': 60, 'min': 60, 'mins': 60, 'minute': 60, 'minutes': 60,
        'h': 3600, 'hr': 3600, 'hrs': 3600, 'hour': 3600, 'hours': 3600,
        'd': 86400, 'day': 86400, 'days': 86400,
        'w': 604800, 'wk': 604800, 'wks': 604800, 'week': 604800, 'weeks': 604800,
    }
    
    total = 0.0
    
    pattern = r'(\d+(?:\.\d+)?)\s*([a-zA-Z]+)'
    matches = re.findall(pattern, s)
    
    if not matches:
        raise ValueError(f"No parseable units in: {string}")
    
    for value_str, unit in matches:
        value = float(value_str)
        if value < 0:
            raise ValueError("Negative duration")
        unit_lower = unit.lower()
        if unit_lower not in unit_map:
            raise ValueError(f"Unknown unit: {unit}")
        total += value * unit_map[unit_lower]
    
    return total


def human_date(timestamp: Union[int, float, str], reference: Optional[Union[int, float, str]] = None) -> str:
    from datetime import timezone
    
    ts = _to_timestamp(timestamp)
    ref = _to_timestamp(reference) if reference is not None else ts
    
    ts_dt = datetime.fromtimestamp(ts, tz=timezone.utc)
    ref_dt = datetime.fromtimestamp(ref, tz=timezone.utc)
    
    ts_date = ts_dt.date()
    ref_date = ref_dt.date()
    
    diff_days = (ts_date - ref_date).days
    
    if diff_days == 0:
        return "Today"
    elif diff_days == -1:
        return "Yesterday"
    elif diff_days == 1:
        return "Tomorrow"
    elif -6 <= diff_days <= -2:
        weekdays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
        return f"Last {weekdays[ts_date.weekday()]}"
    elif 2 <= diff_days <= 6:
        weekdays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
        return f"This {weekdays[ts_date.weekday()]}"
    else:
        months = ["January", "February", "March", "April", "May", "June",
                  "July", "August", "September", "October", "November", "December"]
        month_name = months[ts_date.month - 1]
        if ts_date.year == ref_date.year:
            return f"{month_name} {ts_date.day}"
        else:
            return f"{month_name} {ts_date.day}, {ts_date.year}"


def date_range(start: Union[int, float, str], end: Union[int, float, str]) -> str:
    from datetime import timezone
    
    start_ts = _to_timestamp(start)
    end_ts = _to_timestamp(end)
    
    if start_ts > end_ts:
        start_ts, end_ts = end_ts, start_ts
    
    start_dt = datetime.fromtimestamp(start_ts, tz=timezone.utc)
    end_dt = datetime.fromtimestamp(end_ts, tz=timezone.utc)
    
    start_date = start_dt.date()
    end_date = end_dt.date()
    
    months = ["January", "February", "March", "April", "May", "June",
              "July", "August", "September", "October", "November", "December"]
    
    start_month = months[start_date.month - 1]
    end_month = months[end_date.month - 1]
    
    if start_date == end_date:
        return f"{start_month} {start_date.day}, {start_date.year}"
    elif start_date.year == end_date.year and start_date.month == end_date.month:
        return f"{start_month} {start_date.day}–{end_date.day}, {start_date.year}"
    elif start_date.year == end_date.year:
        return f"{start_month} {start_date.day} – {end_month} {end_date.day}, {start_date.year}"
    else:
        return f"{start_month} {start_date.day}, {start_date.year} – {end_month} {end_date.day}, {end_date.year}"
